### ISSUE TITLE
data-vs-variable derivative is miscalculated

### ISSUE THREAD

**Issue by jsberg-bnl** [bmad-sim/bmad-ecosystem#227]
Ridiculously complicated example in `/cdat/tem/jsberg/220603a`. Run `tao`, then `call 220603a-bug.tao`. The output should demonstrate that tao is miscalculating the derivative (end of that output follows below). The `deriv` command claims that the nonzero derivatives are with respect to variables `ir10_41[8]` and `ir10_41[9]`, but you can see from the definition of `ir10_41.sym[1]` that the derivative should be nonzero with respect to variables `ir10_41[7]` and `ir10_41[8]` instead. And this is manifested in the optimizer behaving crazily, so it's not just a misprint in `show derivative`. And just to confuse things, there are virturally identical data/variables elsewhere (see the similar `ir4` variables and data) that behave as expected.
```
Tao: sho deriv
[INFO] tao_dmodel_dvar_calc:
    Remaking dModel_dVar derivative matrix.
    This may take a while...

Data                          Variable                         Derivative   ix_dat  ix_var
25@ir10_41.sym[1]             ir10_41[1]                       0.00000E+00     1     1
25@ir10_41.sym[1]             ir10_41[2]                       0.00000E+00     1     2
25@ir10_41.sym[1]             ir10_41[3]                       0.00000E+00     1     3
25@ir10_41.sym[1]             ir10_41[4]                       0.00000E+00     1     4
25@ir10_41.sym[1]             ir10_41[5]                       0.00000E+00     1     5
25@ir10_41.sym[1]             ir10_41[6]                       0.00000E+00     1     6
25@ir10_41.sym[1]             ir10_41[7]                       0.00000E+00     1     7
25@ir10_41.sym[1]             ir10_41[8]                       1.00000E+00     1     8
25@ir10_41.sym[1]             ir10_41[9]                       1.00000E+00     1     9
25@ir10_41.sym[1]             ir10_41[10]                      0.00000E+00     1    10
25@ir10_41.sym[1]             ir10_41[11]                      0.00000E+00     1    11
25@ir10_41.sym[1]             ir10_41[12]                      0.00000E+00     1    12
25@ir10_41.sym[1]             ir10_41[13]                      0.00000E+00     1    13
25@ir10_41.sym[1]             ir10_41[14]                      0.00000E+00     1    14
25@ir10_41.sym[1]             ir10_41[15]                      0.00000E+00     1    15
25@ir10_41.sym[1]             ir10_41[16]                      0.00000E+00     1    16
25@ir10_41.sym[1]             ir10_41[17]                      0.00000E+00     1    17
25@ir10_41.sym[1]             ir10_41[18]                      0.00000E+00     1    18
25@ir10_41.sym[1]             ir10_41[19]                      0.00000E+00     1    19
25@ir10_41.sym[1]             ir10_41[20]                      0.00000E+00     1    20
Tao: sho dat ir10_41.sym[1,2]
Universe: 25  of: 30
Data name: ir10_41.sym
                                                                                                 |   Useit
                  Ref_Ele   Start_Ele   Ele                 Meas           Model          Design | Opt  Plot
    1  25@data::ir10_41.low_lim[7] - 25@data::ir10_41.high_lim[8] <target>
                                                   0.0000000E+00  -6.3555925E-02   0.0000000E+00     T     F
    2  25@data::ir10_41.high_lim[8] - 25@data::ir10_41.low_lim[1] <target>
                                                   0.0000000E+00   5.8181481E-02   0.0000000E+00     F     F
                  Ref_Ele   Start_Ele   Ele                 Meas           Model          Design | Opt  Plot
                                                                                                 |   Useit
Tao: sho dat ir10_41.low_lim[7,8]
Universe: 25  of: 30
Data name: ir10_41.low_lim
                                                                                                 |   Useit
                  Ref_Ele   Start_Ele   Ele                 Meas           Model          Design | Opt  Plot
    7  var::ir10_41[7]|model   - var::ir10_41[7]|low_lim <target>
                                                   0.0000000E+00   2.3839642E+01   2.3839642E+01     F     F
    8  var::ir10_41[8]|model   - var::ir10_41[8]|low_lim <target>
                                                   0.0000000E+00   5.1209680E+02   5.1209680E+02     F     F
                  Ref_Ele   Start_Ele   Ele                 Meas           Model          Design | Opt  Plot
                                                                                                 |   Useit
Tao: sho dat ir10_41.high_lim[7,8]
Universe: 25  of: 30
Data name: ir10_41.high_lim
                                                                                                 |   Useit
                  Ref_Ele   Start_Ele   Ele                 Meas           Model          Design | Opt  Plot
    7  var::ir10_41[7]|high_lim   - var::ir10_41[7]|model <target>
                                                   0.0000000E+00   3.9216036E+02   3.9216036E+02     F     F
    8  var::ir10_41[8]|high_lim   - var::ir10_41[8]|model <target>
                                                   0.0000000E+00   2.3903198E+01   2.3903198E+01     F     F
                  Ref_Ele   Start_Ele   Ele                 Meas           Model          Design | Opt  Plot
                                                                                                 |   Useit
```

**Comment by DavidSagan** [comment]
Problem is that ir10_41.sym[1] depends upon ir10_41.low_lim[7,8] which are evaluated *after* ir10_41.sym[1]. I have put in error checking in the code for this (Note: This is documented in the manual).

**Comment by jsberg-bnl** [comment]
Yup, that was it, thanks! I should make another pass through the manual for bedtime reading. 
While I have no specific case at this point, I could imagine someone getting sufficiently creative that a dependency problem couldn't get resolved, particular if cross-universe calculations were involved. You might try something like the following:
1. Add an `evaluated` flag to each datum
2. Make a pass through setting each `evaluated` flag to false
3. Make your pass through the data, evaluating anything you can and setting its `evaluated` flag, skipping anything whose dependencies do not have their `evaluated` flag set
4. Repeat the pass through the data until there are no remaining data with `evaluated` still false

As an alternative, when you find unevaluated dependencies, you could recursively evaluate them; that would probably be more efficient but more difficult to code.
Anyhow, I have no use case for needing such a feature and have no specific example in mind, so you can fee free to ignore this or give it a low priority.

**Comment by DavidSagan** [comment]
You should always be able to arrange the data order to avoid the problem. The exception is a circular dependency which is a pathology.

**Comment by DavidSagan** [comment]
Reporting this was useful since the code did need to do error checking. So please report other cases where there is a lack of error checking

**Comment by jsberg-bnl** [comment]
I thought about that, but I think the issue could with universes: 1@d1 -> 2@d2 -> 1@d3. Since you do universes first, this can't be resolved. But arguably sufficiently pathological that it would be unlikely to come up in practice. Though data order then universe order would solve everything I can think of.

**Comment by DavidSagan** [comment]
For datums that use expressions (which is what we are dealing with here) can be assigned to any universe so it should always be possible to arrange things appropriately.

**Comment by jsberg-bnl** [comment]
Good point, I hadn't thought that through (I usually think of expressions as belonging to some universe but yes, that is completely artificial).
